File: 3 Sum/ans.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        i = 0
        ans = []

        while i < len(nums) - 2:
            
            if i > 0 and nums[i] == nums[i - 1]:
                i += 1
                continue

            j = i + 1
            k = len(nums) - 1

            while j < k:
                sum_val = nums[i] + nums[j] + nums[k]
                if sum_val == 0:
                    ans.append([nums[i], nums[j], nums[k]])
                    j += 1
                    k -= 1
                    while j < k and nums[j] == nums[j - 1]:
                        j += 1
                    while j < k and nums[k] == nums[k + 1]:
                        k -= 1
                elif sum_val < 0:
                    j += 1
                else:
                    k -= 1

            i += 1
        return ans

        


        

File: 3 Sum/ans.cpp

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int i =0;
        vector <vector<int>> ans;
        while (i < nums.size()-2){
            if (i>0 && nums[i] == nums[i-1]){
                i++;
                continue;
            }
            int j =i+1 ;
            int k = nums.size()-1;
            while (j < k){
                int sum_val = nums[i] + nums[j] + nums[k];
                if (sum_val == 0){
                    ans.push_back({nums[i], nums[j], nums[k]});
                    j++;
                    k--;
                    while (j<k && nums[j] == nums[j-1]){
                        j++;
                    }
                    while (j<k && nums[k] == nums[k+1]){
                        k--;
                    }
                }
                else if( sum_val < 0){
                    j++;
                }else{
                    k--;
                }


            }
            i++;

        }
        return ans;
        
    }

};
